R/DataMart.R

```
@@ -484,6 +504,26 @@ getCompanySecret <- function()
     return (secret)
 }
 
+#' Gets document secret from the environment. Throws an error if not found.
+#'
+#' @return Document secret token as a string.
+#'
+#' @noRd
+getProjectSecret <- function()
+{
+    secret <- get0("projectSecret", ifnotfound = "")
+    # projectSecret might not have been copied into global projectSecret by a newer R Server, 
+    # but it could have been stored by QServer in user secrets which are copied into userSecrets by older R servers.
+    if (secret == "") {
+        secret <- tryCatch({
+            val <- .GlobalEnv$userSecrets$projectSecret
+            if (is.character(val) && nzchar(val)) val else ""
+        },
+        error = function(e) "")
+    }
```

This is a temporary shortcut that makes possible to work with the modified `flipAPI` sooner, i.e. without the need to wait for an updated R server to be deployed.
One we make the necessary change in R server, this can be removed.

---

This swallows errors. Is this intended? I'm not following the temporary shortcut argument? You can work with `flipAPI` sooner before a separate R server update?

---

The idea of the shortcut is demonstrated in the screenshot above.

As you can see `.GlobalEnv$userSecrets$projectSecret` carries a value of a Guid. This value is set for every RItem in a Displayr Document [here](https://github.com/Displayr/q/blob/04d859e6c6476e75e3d9b9adcd05f603a2e05282/QLib/DataModel/AnalysisTree/RExecution/RCodeExecutionTask.cs#L279-L281). R server already knows to pass company secret and user secrets into the global environment, but not project secret. Since user secrets is very generic by definition, I piggyback on it to pass a secret named `projectSecret` into `flipAPI`.
My thought about the shortcut was that it would be possible for `flipAPI` to get the secret from R server through user secrets. I now realize that if `flipAPI` is not installed in a running R server dynamically, the shortcut won't work.

Do you know whether `flipAPI` is bundled in R server image or is it installed dynamically?
Would `install.packages()` in an RItem work to install a newer package version?